File: Hacktoberfest2023-main/Add Code Here/LeetCode Problems/Longest String Chain.cpp
class Solution {
public:
	bool predecessor(string &prev, string &curr) {
		int n = prev.size(), m = curr.size();
        
        
		if (n+1 != m) {
            return false;
        }

        int i=0, j=0;
		while (j < m) {
			if (prev[i] == curr[j]) {
				i++, j++;
			}
			else j++;
		}

       
		return i==n;
	}
    
    
    static bool compare (string& first, string& second) {
        return first.size() < second.size();
    }
    
	int longestStrChain(vector<string>& words) {
        
        
		sort(words.begin(), words.end(), compare);
        

		vector<int> dp(words.size(), 1);
		int ans = 1; 
    
		for (int curr = 0; curr < words.size(); curr++) {
			for (int prev = 0; prev < curr; prev++) {
				if (predecessor(words[prev], words[curr])) {
                    dp[curr] = max(dp[curr], dp[prev]+1);
                }
			}
		
			ans = max(ans, dp[curr]);
		}
		return ans;
	}
};

File: Add Code Here/Java/Longest String Chain.java
class Solution {
    public int longestStrChain(String[] words) {
        Arrays.sort(words, (s1, s2) -> Integer.compare(s1.length(), s2.length()));
        Map<String, Integer> map = new HashMap<>();
        int ans = 0;
        
        for (String word : words) {
            int longest = 0;
            for (int i = 0; i < word.length(); i++) {
                StringBuilder sub = new StringBuilder(word);
                sub.deleteCharAt(i);
                String subStr = sub.toString();
                longest = Math.max(longest, map.getOrDefault(subStr, 0) + 1);
            }
            map.put(word, longest);
            ans = Math.max(ans, longest);
        }
        
        return ans;
    }
}
